File: src/features/letter-flow-game/engine.ts
// src/features/letter-flow-game/engine.ts
/**
 * @description Game engine for the Letter Flow game.
 * This engine handles loading game levels, generating game boards, and assigning colors.
 * --- It now extends the BaseGameEngine for architectural consistency,
 * but overrides the `loadLevels` method to handle its unique, non-category-based loading logic. ---
 */
import type { Language, Difficulty, GameCategory, GameLevel } from '@/types/game';
import { shuffleArray } from '@/lib/gameUtils';
import { BaseGameEngine } from '@/games/engine/BaseGameEngine';

/**
 * Represents a single cell on the game board
 */
export interface BoardCell {
  x: number;           // X coordinate on the board
  y: number;           // Y coordinate on the board
  letter: string;      // Letter displayed in this cell
  isUsed: boolean;     // Whether this cell has been used in a word path
  color?: string;      // CSS color for visual distinction
}

/**
 * --- This interface is now correctly exported. ---
 * Represents a path that forms a word on the game board
 */
export interface WordPath {
  word: string;        // The word formed by this path
  cells: BoardCell[]; // Array of cells that make up this path
  startIndex: number;  // Starting index of the word in the solution
}

/**
 * Represents a complete level in the Letter Flow game
 */
export interface LetterFlowLevel extends GameLevel {
  id: string;
  difficulty: Difficulty;
  words: string[];
  board: BoardCell[];
  solution: string;
  endpoints: { x: number; y: number; letter: string; color?: string }[];
}

/**
 * Represents a module structure for level data
 */
interface LevelModule {
  default?: { levels?: unknown[] };
}

/**
 * Generates an HSL color with evenly distributed hues for visual distinction.
 */
function hslForIndex(index: number, total: number, saturation = 70, lightness = 50): string {
  if (total <= 0) return `hsl(0, ${saturation}%, ${lightness}%)`;
  const hue = Math.round((index * 360) / total) % 360;
  const adjustedLightness = lightness > 50 ? lightness * 0.9 : lightness * 1.1;
  return `hsl(${hue}, ${saturation}%, ${adjustedLightness}%)`;
}

/**
 * Main game engine class for the Letter Flow game.
 * --- Now extends BaseGameEngine. ---
 */
class LetterFlowGameEngine extends BaseGameEngine<LetterFlowLevel> {
  /**
   * --- This engine has unique logic for board generation within the loading process. ---
   * We override the base `loadLevels` to preserve this specific functionality.
   */
  public async loadLevels(options: {
    language: Language;
    difficulty?: Difficulty;
  }): Promise<LetterFlowLevel[]> {
    const { language, difficulty } = options;
    if (!difficulty) return [];

    try {
      const path = this.getModulePath(language, '' as GameCategory, difficulty);
      const modules = import.meta.glob('/src/data/**/*.json');
      const moduleLoader = modules[path];
      if (!moduleLoader) return [];

      const module = (await moduleLoader()) as LevelModule;
      const levels = module.default?.levels || [];

      return levels
        .map((lvl: unknown): LetterFlowLevel | null => this.validateLevel(lvl, difficulty))
        .filter((l): l is LetterFlowLevel => l !== null);
    } catch (err) {
      console.error(`LetterFlowGameEngine: Failed to load levels for ${language}/${difficulty}.`, err);
      return [this.getErrorLevel()];
    }
  }

  // --- Implementation of abstract methods from BaseGameEngine ---

  protected getGameId(): string {
    return 'letter-flow';
  }

  protected getModulePath(language: Language, _: GameCategory, difficulty?: Difficulty): string {
    return `/src/data/${language}/letter-flow/${difficulty}.json`;
  }

  protected validateLevel(levelData: unknown, difficulty?: Difficulty): LetterFlowLevel | null {
    const lvl = levelData as Record<string, unknown>;
    if (
      lvl && typeof lvl === 'object' &&
      'id' in lvl && 'solutionWord' in lvl && 'gridSize' in lvl && 'endpoints' in lvl
    ) {
      const { width, height } = (lvl as { gridSize: { width: number; height: number } }).gridSize;
      const board: BoardCell[] = Array.from({ length: width * height }, (_, i) => ({
        x: i % width, y: Math.floor(i / width), letter: '', isUsed: false
      }));

      const endpoints = (lvl as { endpoints: { x: number; y: number; letter: string }[] }).endpoints;
      const uniqueLetters = Array.from(new Set(endpoints.map(e => e.letter)));
      const letterToColor = new Map<string, string>();
      uniqueLetters.forEach((letter, idx) => {
        letterToColor.set(letter, hslForIndex(idx, uniqueLetters.length, 72, 48));
      });

      endpoints.forEach((endpoint) => {
        const index = endpoint.y * width + endpoint.x;
        if (index < board.length) {
          board[index] = { ...endpoint, isUsed: false, color: letterToColor.get(endpoint.letter) };
        }
      });

      const typedLvl = lvl as { id: string | number; solutionWord: string; endpoints: { x: number; y: number; letter: string }[] };
      return {
        id: String(typedLvl.id),
        difficulty: difficulty ?? 'easy',
        words: [typedLvl.solutionWord],
        board,
        solution: typedLvl.solutionWord,
        endpoints: typedLvl.endpoints.map(e => ({ ...e, color: letterToColor.get(e.letter) })),
      };
    }
    return null;
  }

  protected getErrorLevel(): LetterFlowLevel {
    return {
      id: 'error',
      difficulty: 'easy',
      words: [],
      board: [],
      solution: '',
      endpoints: [],
    };
  }

  /**
   * --- This public method is now correctly restored to the class. ---
   * Generate a game board from a string of letters. This is used as a fallback.
   */
  public generateBoard(_s: string, _d: Difficulty, _l: Language, baseLetters?: string): BoardCell[] {
    if (!baseLetters) return [];

    // Shuffle the letters for random placement
    const letters = shuffleArray(baseLetters.split(''));
    // Calculate board size (square root of letter count, rounded up)
    const size = Math.ceil(Math.sqrt(letters.length));
    const board: BoardCell[] = [];

    // Create board cells for each position
    for (let y = 0; y < size; y++) {
      for (let x = 0; x < size; x++) {
        const index = y * size + x;
        if (index < letters.length) {
          board.push({
            x,
            y,
            letter: letters[index],
            isUsed: false,
          });
        }
      }
    }

    return board;
  }
}

/**
 * Create and export a singleton instance of the game engine.
 */
export const letterFlowGameEngine = new LetterFlowGameEngine();

File: crates/power-game-letterflow/src/engine.rs
//! Letter Flow Game Engine Core

use crate::models::{BoardCell, LetterFlowLevel, PathPoint, GridSize};
use serde_json::Value;

/// Main engine struct for Letter Flow game
pub struct LetterFlowEngine;

impl LetterFlowEngine {
    /// Validates level data has required fields
    pub fn validate_level_data(level_data: &Value) -> bool {
        level_data.get("board").is_some() && 
        level_data.get("solution").is_some()
    }
    
    /// Extract grid size from level data
    pub fn extract_grid_size(level_data: &Value) -> Option<GridSize> {
        let grid_size = level_data.get("gridSize")?;
        let width = grid_size.get("width")?.as_i64()? as i32;
        let height = grid_size.get("height")?.as_i64()? as i32;
        Some(GridSize { width, height })
    }
}

/// Generates a square game board from a string of letters
/// 
/// Calculates the square root of the letter count, rounding up,
/// to determine the grid size (N x N), then fills the grid row by row.
pub fn generate_board(letters: &str) -> Vec<BoardCell> {
    let chars: Vec<char> = letters.chars().collect();
    
    // Board size is Ceiling(sqrt(letter_count))
    let size = (chars.len() as f64).sqrt().ceil() as i32;
    let mut board = Vec::with_capacity(chars.len());

    for y in 0..size {
        for x in 0..size {
            let index = (y * size + x) as usize;
            if index < chars.len() {
                board.push(BoardCell {
                    x,
                    y,
                    letter: chars[index].to_string(),
                    is_used: false,
                    color: None,
                });
            }
        }
    }

    board
}

/// Generate HSL color for index with even distribution
pub fn hsl_for_index(index: usize, total: usize, saturation: u8, lightness: u8) -> String {
    if total == 0 {
        return format!("hsl(0, {}%, {}%)", saturation, lightness);
    }
    let hue = ((index * 360) / total) % 360;
    let adjusted_lightness = if lightness > 50 { 
        (lightness as f32 * 0.9) as u8 
    } else { 
        (lightness as f32 * 1.1) as u8 
    };
    format!("hsl({}, {}%, {}%)", hue, saturation, adjusted_lightness)
}

/// Assign colors to endpoints based on their letters
pub fn assign_endpoint_colors(endpoints: &[PathPoint]) -> Vec<PathPoint> {
    let unique_letters: Vec<String> = endpoints
        .iter()
        .map(|e| e.letter.clone())
        .collect::<std::collections::HashSet<_>>()
        .into_iter()
        .collect();
    
    let letter_to_color: std::collections::HashMap<String, String> = unique_letters
        .iter()
        .enumerate()
        .map(|(idx, letter)| {
            let color = hsl_for_index(idx, unique_letters.len(), 72, 48);
            (letter.clone(), color)
        })
        .collect();
    
    endpoints
        .iter()
        .map(|ep| PathPoint {
            x: ep.x,
            y: ep.y,
            letter: ep.letter.clone(),
            color: letter_to_color.get(&ep.letter).cloned(),
        })
        .collect()
}

/// Build board from level data with endpoints
pub fn build_board_from_level(level_data: &Value) -> Option<Vec<BoardCell>> {
    let grid_size = LetterFlowEngine::extract_grid_size(level_data)?;
    let width = grid_size.width;
    let height = grid_size.height;
    
    // Initialize empty board
    let mut board: Vec<BoardCell> = (0..width * height)
        .map(|i| BoardCell {
            x: i % width,
            y: i / width,
            letter: String::new(),
            is_used: false,
            color: None,
        })
        .collect();
    
    // Place endpoints
    let endpoints = level_data.get("endpoints")?;
    for endpoint in endpoints.as_array()? {
        let x = endpoint.get("x")?.as_i64()? as i32;
        let y = endpoint.get("y")?.as_i64()? as i32;
        let letter = endpoint.get("letter")?.as_str()?;
        let color = endpoint.get("color").and_then(|c| c.as_str());
        
        let index = (y * width + x) as usize;
        if index < board.len() {
            board[index].letter = letter.to_string();
            board[index].color = color.map(|c| c.to_string());
        }
    }
    
    Some(board)
}

/// Create error level when loading fails
pub fn create_error_level() -> LetterFlowLevel {
    LetterFlowLevel {
        id: "error".to_string(),
        difficulty: "easy".to_string(),
        words: vec![],
        board: vec![],
        solution: String::new(),
        endpoints: vec![],
    }
}

/// Parse and validate level from JSON
pub fn parse_level(level_json: &str) -> Option<LetterFlowLevel> {
    let value: Value = serde_json::from_str(level_json).ok()?;
    
    let id = value.get("id")?.as_str()?.to_string();
    let difficulty = value.get("difficulty")
        .and_then(|d| d.as_str())
        .unwrap_or("easy")
        .to_string();
    let solution = value.get("solutionWord")
        .or_else(|| value.get("solution"))?
        .as_str()?
        .to_string();
    
    let board = build_board_from_level(&value)?;
    
    // Extract endpoints
    let endpoints: Vec<PathPoint> = value
        .get("endpoints")?
        .as_array()?
        .iter()
        .filter_map(|ep| {
            Some(PathPoint {
                x: ep.get("x")?.as_i64()? as i32,
                y: ep.get("y")?.as_i64()? as i32,
                letter: ep.get("letter")?.as_str()?.to_string(),
                color: ep.get("color").and_then(|c| c.as_str()).map(|c| c.to_string()),
            })
        })
        .collect();
    
    let words = vec![solution.clone()];
    
    Some(LetterFlowLevel {
        id,
        difficulty,
        words,
        board,
        solution,
        endpoints: assign_endpoint_colors(&endpoints),
    })
}
